Foxtrot/Driver/Rewriting/EmitAsyncClosure.cs

```diff
@@ -132,21 +248,45 @@ public EmitAsyncClosure(Method from, Rewriter rewriter)
             this.func2Type = new Cache<TypeNode>(() =>
                 HelperMethods.FindType(SystemTypes.SystemAssembly, StandardIds.System, Identifier.For("Func`2")));
 
-            if (from.IsGeneric)
+            // Should distinguish between generic enclosing method and non-generic method in enclosing type.
+            // In both cases generated closure should be generic.
+            Func<Method, TypeNodeList> getGenericTypesFrom = method =>
+            {
+                if (method.IsGeneric)
+                {
+                    return from.TemplateParameters;
+                }
+                
+                if (method.DeclaringType.IsGeneric)
+                {
+                    return method.DeclaringType.TemplateParameters;
+                }
+
+                return null;
+            };
+
+            var enclosingTemplateParameters = getGenericTypesFrom(from);
```

Comment:
You're right. I thought to extract method but left with "local" function. Will extract function.